include/particles/particle_xmpm.tcc

```diff
@@ -0,0 +1,266 @@
+//! Construct a particle with id and coordinates
+template <unsigned Tdim>
+mpm::ParticleXMPM<Tdim>::ParticleXMPM(Index id, const VectorDim& coord)
+    : mpm::Particle<Tdim>(id, coord) {
+
+  this->initialise();
+  // Logger
+  std::string logger =
+      "particlexmpm" + std::to_string(Tdim) + "d::" + std::to_string(id);
+  console_ = std::make_unique<spdlog::logger>(logger, mpm::stdout_sink);
+}
+
+//! Construct a particle with id, coordinates and status
+template <unsigned Tdim>
+mpm::ParticleXMPM<Tdim>::ParticleXMPM(Index id, const VectorDim& coord,
+                                      bool status)
+    : mpm::Particle<Tdim>(id, coord, status) {
+  this->initialise();
+  //! Logger
+  std::string logger =
+      "particlexmpm" + std::to_string(Tdim) + "d::" + std::to_string(id);
+  console_ = std::make_unique<spdlog::logger>(logger, mpm::stdout_sink);
+}
+
+// Initialise particle properties
+template <unsigned Tdim>
+void mpm::ParticleXMPM<Tdim>::initialise() {
+  levelset_phi_ = 0.;
+}
+
+//! Map particle mass and momentum to nodes
+template <unsigned Tdim>
+void mpm::ParticleXMPM<Tdim>::map_mass_momentum_to_nodes() noexcept {
+  // Check if particle mass is set
+  assert(mass_ != std::numeric_limits<double>::max());
+
+  // Map mass and momentum to nodes
+  for (unsigned i = 0; i < nodes_.size(); ++i) {
+    nodes_[i]->update_mass(true, mpm::ParticlePhase::Solid,
+                           mass_ * shapefn_[i]);
+    nodes_[i]->update_momentum(true, mpm::ParticlePhase::Solid,
+                               mass_ * shapefn_[i] * velocity_);
+    if (nodes_[i]->discontinuity_enrich()) {
+      // Unit 1x1 Eigen matrix to be used with scalar quantities
+      Eigen::Matrix<double, 1, 1> nodal_mass;
+      nodal_mass(0, 0) = sgn(levelset_phi_) * mass_ * shapefn_[i];
+      // Map enriched mass and momentum to nodes
+      nodes_[i]->update_discontinuity_property(true, "mass_enrich", nodal_mass,
+                                               0, 1);
+      nodes_[i]->update_discontinuity_property(true, "momenta_enrich",
+                                               velocity_ * nodal_mass, 0, Tdim);
+    }
+  }
+}
+
+// Compute strain rate of the particle
+template <>
+inline Eigen::Matrix<double, 6, 1> mpm::ParticleXMPM<3>::compute_strain_rate(
```

Comment:
Don't we need to treat the 2D version of discontinuity too? In 2D compute_strain_rate will call the base class, which is the one written in particle.tcc

Reply:
Finish 3D first and i will enrich the 2D part.